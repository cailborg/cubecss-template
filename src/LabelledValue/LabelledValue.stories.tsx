import { Story, Meta } from "@storybook/react/types-6-0";
import { LabelledValue, LabelledValueProps } from "./LabelledValue";

export default {
  title: "Components/LabelledValue",
  component: LabelledValue,
  argTypes: {
    label: { control: "text" },
    value: { control: "text" },
    description: { control: "text" },
    secondaryValue: { control: "text" },
  }
} as Meta;

export const Horizontal: Story<LabelledValueProps> = (args) => <LabelledValue {...args}></LabelledValue>;
Horizontal.args = { label:"Label", variant: "horizontal", value: "value", description: "description", secondaryValue: "secondaryValue", emphasized: false };

export const Vertical: Story<LabelledValueProps> = (args) => <LabelledValue {...args}></LabelledValue>;
Vertical.args = { label:"Label", variant: "vertical", value: "value", description: "description", secondaryValue: "secondaryValue", emphasized: false };

