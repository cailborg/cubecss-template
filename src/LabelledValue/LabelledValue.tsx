import React, { useState } from "react";
import styled from "@emotion/styled";

type LabelledValueVariantType = "vertical" | "horizontal";
type LabelledValueSizeType = "large" | "small";

export interface LabelledValueProps {
    variant: LabelledValueVariantType;
    size?: LabelledValueSizeType;
    label: String;
    value: String;
    description?: String;
    secondaryValue?: String;
    emphasized?: boolean;
}

const Vertical = styled.div`
    display: flex;
    flex-direction: column;
`;

const Description = styled.span`
    font-family: var(--lp-fonts-caption);
    font-weight: var(--lp-font-weight-caption);
    font-size: var(--lp-font-size-caption);
    line-height: var(--lp-line-height-caption);
    letter-spacing: var(--lp-letter-spacing-caption);
    color: rgba(var(--lp-colors-neutral-600), 1);
`;

const Horizontal = styled.div`
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    & > span {
        display: inline-block;
        position: relative;
        overflow-wrap: break-word;
        word-wrap: break-word;
        word-break: break-all;
        word-break: break-word;
        -ms-hyphens: auto;
        -moz-hyphens: auto;
        -webkit-hyphens: auto;
        hyphens: auto;
    }
`;

const LabelValue = styled.span<{ emphasized: boolean; size: string }>`
    font-family: var(--lp-fonts-body-3);
    font-weight: ${(props) =>
        props.emphasized
            ? "var(--lp-font-weight-body-3-bold)"
            : "var(--lp-font-weight-body-3)"};
    font-size: ${(props) =>
        props.size === "small"
            ? "var(--lp-font-size-body-3)"
            : "var(--lp-font-size-body-2)"};
    line-height: var(--lp-line-height-body-3);
    letter-spacing: var(--lp-letter-spacing-body-3);
    color: rgba(var(--lp-colors-neutral-800), 1);
`;

const SecondaryValue = styled.span`
    font-family: var(--lp-fonts-caption);
    font-weight: var(--lp-font-weight-caption);
    font-size: var(--lp-font-size-caption);
    line-height: var(--lp-line-height-caption);
    letter-spacing: var(--lp-letter-spacing-caption);
    color: rgba(var(--lp-colors-neutral-600), 1);
    display: inline-block;
    justify-content: flex-end;
    text-align: right;
`;

export const LabelledValue: React.FC<LabelledValueProps> = ({
    variant,
    size = "small",
    label,
    value,
    description,
    secondaryValue,
    emphasized = false,
}) => {
    const [ariaId] = useState(Date.now().toString());
    if (variant === "vertical") {
        return (
            <Vertical aria-live="polite" aria-atomic="true">
                <Description id={ariaId}>{label}</Description>
                <LabelValue
                    size={size}
                    emphasized={emphasized}
                    aria-labelledby={ariaId}
                >
                    {value}
                </LabelValue>
                {description && <Description>{description}</Description>}
            </Vertical>
        );
    }

    return (
        <div aria-live="polite" aria-atomic="true">
            <Horizontal>
                <LabelValue size={size} emphasized={emphasized} id={ariaId}>
                    {label}
                </LabelValue>
                <LabelValue
                    size={size}
                    emphasized={emphasized}
                    aria-labelledby={ariaId}
                >
                    {value}
                </LabelValue>
            </Horizontal>
            <Horizontal>
                {description && <Description>{description}</Description>}
                {secondaryValue && (
                    <SecondaryValue>{secondaryValue}</SecondaryValue>
                )}
            </Horizontal>
        </div>
    );
};
