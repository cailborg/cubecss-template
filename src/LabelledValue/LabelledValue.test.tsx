import React from "react";
import { LabelledValue } from "./LabelledValue";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer.create(<LabelledValue variant="vertical" label="Label" value="value" description="description" secondaryValue="secondaryValue" emphasized={false}></LabelledValue>).toJSON();
  expect(tree).toMatchSnapshot();
});
