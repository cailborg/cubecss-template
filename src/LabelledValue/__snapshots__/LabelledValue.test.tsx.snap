// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  aria-atomic="true"
  aria-live="polite"
  className="css-1ohf0ui"
>
  <span
    className="css-di4ebj"
    id="1618199163859"
  >
    Label
  </span>
  <span
    aria-labelledby="1618199163859"
    className="css-112hi8s"
    size="small"
  >
    value
  </span>
  <span
    className="css-di4ebj"
  >
    description
  </span>
</div>
`;
